### ISSUE TITLE
[Feature Request] MaxCP in Pokemon Transfer Filter

### ISSUE THREAD

**Issue by CreativiTimothy** [NoxxDev/NecroBot#654]
KeepMinCP is great overall, but occasionally, we need KeepMaxCP. Recently, the BubbleStrat to train gyms to lvl 10 super fast requires 10 cp diglett and 20cp horsea. In order to get one of that low CP range, we need a KeepMaxCP. Could someone please add it? Thanks!

**Comment by fedyay** [comment]
You can specialize keep config for each pokemon, for example:
"PokemonsTransferFilter": {
    "Diglet": {
      "KeepMinCp": 10,
      "KeepMinLvl": 1,
      "UseKeepMinLvl": true,
      "KeepMinIvPercentage": 1.0,
      "KeepMinDuplicatePokemon": 10,
      "Moves": [],
      "DeprecatedMoves": null,
      "KeepMinOperator": "and",
      "MovesOperator": "or"
    },
And make high priority for certain pokemon in HumanSniping section.

**Comment by mo0ojava** [comment]
That snippet will cause NB to crash. Its "Diglett". Otherwise good tip!

**Comment by CreativiTimothy** [comment]
＠ｆｅｄｙａｙ　Ｉ　ａｌｒｅａｄｙ　ｋｎｏｗ．　Ｔｈｅ　ｃｏｎｆｉｇ　ｙｏｕ　ｊｕｓｔ　ｐｏｓｔｅｄ　ｗｉｌｌ　ｋｅｅｐ　ａｌｌ　ｄｉｇｌｅｔｔｓ　１０ｃｐ　ａｎｄ　ｕｐ，　ｗｈｉｃｈ　ｉｓ　ｎｏｔ　ｗｈａｔ　ｗｅ　ｗａｎｔ．　Ｗｅ　ｗａｎｔ　ａ　１０ｃｐ　ｄｉｇｌｅｔｔ，　ｎｏｔ　ａ　１１ｃｐ　ｏｒ　２００ｃｐ．　Ｔｈａｔ’ｓ　ｗｈｙ　ｗｅ　ｎｅｅｄ　ＭａｘＣＰ　／　Ｍａｘ　Ｌｖｌ

**Comment by fedyay** [comment]
@mo0ojava yep, thanks! I don't use this code and wrote it on the run.

@CreativiTimothy Yeah, you are right. Btw, are you fighting for gyms with the same account you bot on?

**Comment by CreativiTimothy** [comment]
@fedyay Yeah, but I only train gyms of my own color -- I don't attack other gyms generally. Just want to get my 100 coins. The reason I'm botting is cause I don't want to waste time sitting at pokestops catching 1000 Pidgeys. I need time to study Japanese and Music Theory at my community college

**Comment by fedyay** [comment]
@CreativiTimothy No problem man, I was just asking because in the discussions where I participate at people are convinced that using gyms leads directly to permanent ban. Just tripplechecking).

**Comment by Silph-Road** [comment]
A more appropriate change would be to make a setting that keeps Pokemons over set CP and/or under set CP. Example, keep diglett over 1000 and under 10 

CP > 1000 = Keep 
AND 
CP < 10 = Keep

We can then add the regular operators and switches.
(crappy IV is also helpful as they will be even weaker and easier to train up gym)

So Something like:
```
"PokemonsTransferFilter": {
"Diglett": {
"KeepOverCp": 1000,
"KeepOverLvl": 30,
"UseKeepOverLvl": true,
"KeepOverIvPercentage": 90.0,
"KeepOverOperator": "and",

"KeepUnderCp": 10,
"KeepUnderLvl": 1,
"UseUnderLvl": true,
"KeepUnderIvPercentage": 10.0,
"KeepUnderOperator": "and",
"UseUnderCPandLVLFilter": true,

"KeepMinDuplicatePokemon": 10,
"Moves": [],
"DeprecatedMoves": null,
"MovesOperator": "or"
},
```

-->Adding a "KeepMaxCP:10" would make you loose very strong pokemons for no reason at all.
**Plus you would have to favorite all your 10CP+ pokemons in order to stop them from being transferred.**

Thanks.
@CreativiTimothy 
@fedyay 
@mo0ojava 
@NoxxDev

**Comment by Silph-Road** [comment]
[For those who don't know what is bubble strat](https://pokemongo.gamepress.gg/bubblestrat)

**Comment by Silph-Road** [comment]
@NoxxDev I know this will require some coding, but look at the battles!!
https://pokemongo.gamepress.gg/bubblestrat-extensions

Thanks.

**Comment by CreativiTimothy** [comment]
@fedyay That's only if you act really obvious. If you're lvl 38 and take over every gym in your area, people are going to report. If you're lvl 30 or something which looks mostly legit, and you only put 1 Lapras and 1 Snorlax instead of 10 Snorlaxes in every gym, then you look very legit.

**Comment by Silph-Road** [comment]
@CreativiTimothy 
"Do you mean < 1000 and >10 CP?"
No, 
Your above mentioned format will catch anything less than "<" 1000 and more than ">" 10

cuz again.... > 10 means more than 10... (no limit, it could be 10, 20, 30, 40, 50, 1000 etc.)

we need exactly 10 or less than a set CP, like < 20
<10 means <=10.... based on the currently implemented operations in the bot:
For example:
"KeepOverCp": 1000, means it will keep 1000 and anything more. It does mean it will keep 1001 but not 1000. 

Thus, 
<10 will keep anything = 10. and theoretically under, but, obviously nothing is under 10CP.

Its use is apparent when you use it as:
<20 will keep anything 20 and under, including 10

**Comment by CreativiTimothy** [comment]
@Silph-Road I'm a straight A+ student don't worry about explaining.  I believe you just made a typo. You had put <10CP and now I noticed you edited to <= 10CP. That makes more sense. It will keep every 10 CP Diglett

**Comment by Silph-Road** [comment]
There was no typo :-) and everyone knows there is no 9 CP pokemon lol

I had to make the change so that "a straight A+ student" can understand that, the bot code interprets < as <=


and I hope you can understand what you typed, so I had to explain..

@CreativiTimothy 

> Do you mean < 1000 and >10 CP?

which means... 
less than "<" 1000 and more than ">" 10

unless that is what you want it to catch, based on what you stated??.... pokemons between 10 and 1000??

I can further go into analysis of your A+ reasoning...

> Like you said, there should be an option called "Enable MaxCP / max IV: true". BY default, it will be disabled, so all your strong pokemon will be kept, but the user will need to set it to true if they want to keep weak IV CP diglettts

Analysis:
"Enable MaxCP / max IV: true" will **eliminate** all pokemons over CP/IV after it is enabled, unless the user goes and favorites every single pokemon of that species over 10 CP <--- kinda pointless and not usefull eh?... over time those favorited pokemons will become weak as you level up and the user will have to go back and unfavorite them to be able to transfer them...

Final words...
Over CP/IV and Under CP/IV is more logical as mentioned above.
